docs/howto/configure/auth-management.md

```diff
@@ -53,4 +55,107 @@ So we want to manage authentication by:
+
+2. **Transfer the OAuth App to the `2i2c-org` GitHub account.**
+   By default, OAuth apps are created under your GitHub account.
+   However, it will be a point of failure if a hub's app is only accessible by a single member of the 2i2c Engineering team.
+   Therefore, please ensure you transfer it to the `2i2c-org` GitHub org by following [GitHub's documentation](https://docs.github.com/en/developers/apps/managing-oauth-apps/transferring-ownership-of-an-oauth-app).
+
+3. **Create or update the appropriate secret config file under `secrets/config/hubs/*.cluster.yaml`.**
+   You should add the following config to this file, pasting in the client ID and secret you generated in step 1.
+
+    ```yaml
+    hubs:
+    - name: HUB_NAME
+      config:
+        jupyterhub:
+          hub:
+            config:
+              GitHubOAuthenticator:
+                client_id: CLIENT_ID
+                client_secret: CLIENT_SECRET
+    ```
+
+    ```{note}
+    Add the `basehub` key between `config` and `jupyterhub` for `daskhub` deployments.
+    ```
+
+    ```{admonition}
+    Make sure this is encrypted with `sops` before committing it to the repository!
+
+    `sops -i -e secrets/config/hub/*.cluster.yaml`
+    ```
+
+4. **Edit the non-secret config under `config/hubs`.**
+   You should make sure the matching hub config takes one of the following forms.
+
+   To authenticate against a GitHub organisation:
+
+    ```yaml
+    hubs:
+    - name: HUB_NAME
+      auth0:
+        enabled: false
+        connection: ""  # This key is still required by the schema
+      ... # Other config
+      config:
+        jupyterhub:
+          hub:
+            config:
+              JupyterHub:
+                authenticator_class: github
+              GitHubOAuthenticator:
+                oauth_callback_url: https://{{ HUB_DOMAIN }}/hub/oauth_callback
+                allowed_organizations:
+                  - 2i2c-org
+                  - ORG_NAME
+                scope:
+                  - read:user
+            extraConfig:
+              06-custom-authenticator: ""  # Required to override our overrides...
+    ```
+
+   To authenticate against a GitHub Team:
+
+    ```yaml
+    hubs:
+    - name: HUB_NAME
+      auth0:
+        enabled: false
+        connection: ""  # This key is still required by the schema
+      ... # Other config
+      config:
+        jupyterhub:
+          hub:
+            config:
+              JupyterHub:
+                authenticator_class: github
+              GitHubOAuthenticator:
+                oauth_callback_url: https://{{ HUB_DOMAIN }}/hub/oauth_callback
+                allowed_organizations:
+                  - 2i2c-org:tech-team
+                  - ORG_NAME:TEAM_NAME
+                scope:
+                  - read:org
+            extraConfig:
+              06-custom-authenticator: ""  # Required to override our overrides...
```

Comment:
I needed to override this in our values, otherwise the JupyterHub returned a 500 error when you clicked on the "Login" button and the hub logs showed:

```
[E 2021-09-23 10:41:12.961 JupyterHub web:1789] Uncaught exception GET /hub/oauth_login?next=%2Fhub%2F (10.128.0.21)
    HTTPServerRequest(protocol='https', host='staging.pangeo.2i2c.cloud', method='GET', uri='/hub/oauth_login?next=%2Fhub%2F', version='HTTP/1.1', remote_ip='10.128.0.21')
    Traceback (most recent call last):
      File "/usr/local/lib/python3.8/dist-packages/tornado/web.py", line 1702, in _execute
        result = method(*self.path_args, **self.path_kwargs)
      File "/usr/local/lib/python3.8/dist-packages/oauthenticator/oauth2.py", line 115, in get
        self.authorize_redirect(
      File "/usr/local/lib/python3.8/dist-packages/tornado/auth.py", line 585, in authorize_redirect
        url = self._OAUTH_AUTHORIZE_URL  # type: ignore
      File "/usr/local/lib/python3.8/dist-packages/oauthenticator/oauth2.py", line 72, in _OAUTH_AUTHORIZE_URL
        return self.authenticator.authorize_url
      File "/usr/local/lib/python3.8/dist-packages/traitlets/traitlets.py", line 575, in __get__
        return self.get(obj, cls)
      File "/usr/local/lib/python3.8/dist-packages/traitlets/traitlets.py", line 538, in get
        default = obj.trait_defaults(self.name)
      File "/usr/local/lib/python3.8/dist-packages/traitlets/traitlets.py", line 1578, in trait_defaults
        return self._get_trait_default_generator(names[0])(self)
      File "/usr/local/lib/python3.8/dist-packages/traitlets/traitlets.py", line 975, in __call__
        return self.func(*args, **kwargs)
      File "/usr/local/lib/python3.8/dist-packages/oauthenticator/auth0.py", line 69, in _authorize_url_default
        return "https://%s.auth0.com/authorize" % self.auth0_subdomain
      File "/usr/local/lib/python3.8/dist-packages/traitlets/traitlets.py", line 575, in __get__
        return self.get(obj, cls)
      File "/usr/local/lib/python3.8/dist-packages/traitlets/traitlets.py", line 538, in get
        default = obj.trait_defaults(self.name)
      File "/usr/local/lib/python3.8/dist-packages/traitlets/traitlets.py", line 1578, in trait_defaults
        return self._get_trait_default_generator(names[0])(self)
      File "/usr/local/lib/python3.8/dist-packages/traitlets/traitlets.py", line 975, in __call__
        return self.func(*args, **kwargs)
      File "/usr/local/lib/python3.8/dist-packages/oauthenticator/auth0.py", line 52, in _auth0_subdomain_default
        raise ValueError(
    ValueError: Please specify $AUTH0_SUBDOMAIN env or CustomOAuthenticator.auth0_subdomain config
```

If there's a more elegant way to do this, I'd be happy to hear suggestions.

Reply:
We currently have auth0 set as the authenticator unconditionally here: https://github.com/2i2c-org/pilot-hubs/blob/8ab995af71cd3c21481176061c51cbc0647f8b60/hub-templates/basehub/values.yaml#L216.

We then set it unconditionally *again* in https://github.com/2i2c-org/pilot-hubs/blob/8ab995af71cd3c21481176061c51cbc0647f8b60/hub-templates/basehub/values.yaml#L320.

Let's unset both of those, and set it conditionally here instead: https://github.com/2i2c-org/pilot-hubs/blob/8ab995af71cd3c21481176061c51cbc0647f8b60/deployer/hub.py#L355

Reply:
I gave this a shot in commit ffb4b95 but please let me know if I got it right. I was a bit confused about the mixture of `oauthenticator.auth0.Auth0OAuthenticator` and `CustomOAuthenticator`.

Reply:
Also if I understand correctly, each hub in every `*.cluster.yaml` file will need to be updated to have `JupyterHub.authenticator_class` explicitly specified now

Reply:
From discussion in [this thread](https://github.com/2i2c-org/pilot-hubs/pull/706/#discussion_r717712945) we agreed that `authenticaotr_class` should be `CustomOAuthenticator` instead. Change was pushed [here](https://github.com/2i2c-org/pilot-hubs/pull/706/commits/b1d6aac999bc8d51890da7a99f0f6aab2ea18af6)